doc: improve repository documentation (#5565)
python-poetry/poetry: docs/repositories.md

### Private Repository Example

#### Installing from private package sources
By default, Poetry discovers and installs packages from [PyPI](https://pypi.org). But, you want to
install a dependency to your project for a [simple API repository](#simple-api-repository)? Let's
do it.

First, [configure](#project-configuration) the [package source](#package-source) repository to your
project.

```bash
poetry source add foo https://pypi.example.org/simple/
```

Then, assuming the repository requires authentication, configure credentials for it.

```bash
poetry config http-basic.foo username password
```

Once this is done, you can add dependencies to your project from this source.

```bash
poetry add --source foo private-package
```

#### Publishing to a private repository

Great, now all that is left is to publish your package. Assuming you'd want to share it privately
with your team, you can configure the
[Upload API](https://warehouse.pypa.io/api-reference/legacy.html#upload-api) endpoint for your
[publishable repository](#publishable-repository).

```bash
poetry config repositories.foo https://pypi.example.org/legacy/
```

{{% note %}}

> If you need to use a different credential to your [package source](#package-sources), then it is

Reviewer: ```suggestion
If you need to use a different credential for your [package source](#package-sources), then it is
```

If your package source requires [credentials](#configuring-credentials) or
[certificates](#certificates), please refer to the relevant sections below.

{{% /note %}}

#### Default Package Source

By default, Poetry configures [PyPI](https://pypi.org) as the default package source for your
project. You can alter this behaviour and exclusively look up packages only from the configured
package sources by adding a **single** source with `default = true`.

```bash
poetry source add --default foo https://foo.bar/simple/
```

{{% warning %}}

Configuring a custom package source as default, will effectively disable [PyPI](https://pypi.org)
as a package source for your project.

{{% /warning %}}

#### Secondary Package Sources

If package sources are configured as secondary, all it means is that these will be given a lower
priority when selecting compatible package distribution that also exists in your default package
source.

You can configure a package source as a secondary source with `secondary = true` in your package
source configuration.

```bash
poetry source add --secondary foo https://foo.bar/simple/
```

There can be more than one secondary package source.

{{% note %}}

> Package sources, even when configured as secondary, will be searched when Poetry looks up packages
> despite the package having been found in the default source or another configured source.

Reviewer: ```suggestion
All package sources (including secondary sources) will be searched during the package lookup process. These network requests will occur for all sources, regardless of if the package is found at one or more sources.
```

Author: Much better

> If you wish to avoid this, you may explicitly specify which source too look in for a particular

Reviewer: ```suggestion
If you wish to avoid this, you may explicitly specify which source to search in for a particular
```

package.

```bash
poetry add --source pypi httpx
```

{{% /note %}}

### Supported Package Sources

#### Python Package Index (PyPI)

Poetry interacts with [PyPI](https://pypi.org) via its
[JSON API](https://warehouse.pypa.io/api-reference/json.html). This is used to retrieve a requested

> package's versions, metadata, files etc.

Reviewer: ```suggestion
package's versions, metadata, files, etc.
```

{{% note %}}

> If the the package's published metadata is invalid, Poetry will download available bdist/sdist to

Reviewer: ```suggestion
If the the package's published metadata is invalid, Poetry will download the available bdist/sdist to
```

inspect it locally to identify the relevant metadata.

{{% /note %}}

If you want to explicitly select a package from [PyPI](https://pypi.org) you can use the `--source`
option with the [`add`]({{< relref "cli#add" >}}) command, like shown below.

```bash
poetry add --source pypi httpx@^0.22.0
```

This will generate the following configuration snippet in your `pyproject.toml` file.

```toml
httpx = {version = "^0.22.0", source = "pypi"}
```

{{% warning %}}

If any source within a project is configured with `default = true`, The implicit `pypi` source will
be disabled and not used for any packages.

{{% /warning %}}

#### Simple API Repository

Poetry can fetch and install package dependencies from public or private custom repositories that

> implement the simple API repository as described in [PEP 503](https://peps.python.org/pep-0503/).

Reviewer: ```suggestion
implement the simple repository API as described in [PEP 503](https://peps.python.org/pep-0503/).
```

Author: This one is as intended as it is "simple API" repository.

Reviewer: I do think that's a lot less clear, and it doesn't match the title of the PEP. At the very least "simple API repository specification" would make it more clear.

poetry config certificates.foo.client-cert /path/to/client.pem
```

## Caches

Poetry employs multiple caches for package sources in order to improve user experience and avoid duplicate network
requests.

The first level cache is a [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control)

> cache for almost all HTTP requests.

Reviewer: ```suggestion
header for almost all HTTP requests.
```

Author: This should be "based cache" as we are using the cache control package with the requests session.

Reviewer: Ah, then linking to docs for the header is a bit misleading.

> cache for almost all HTTP requests.

Author: ```suggestion
header based cache for almost all HTTP requests.
```
